src/dynamic_honey_badger.rs

```diff
@@ -0,0 +1,450 @@
+                self.vote_counts.remove(&old_change);
+            }
+        }
+        *self.vote_counts.entry(change).or_insert(0) += 1;
+    }
+
+    /// Returns the change that currently has a majority of votes, if any.
+    fn current_majority(&self) -> Option<&Change<NodeUid>> {
+        self.vote_counts
+            .iter()
+            .find(|&(_, count)| count * 2 > self.netinfo.num_nodes())
+            .map(|(change, _)| change)
+    }
+}
+
+/// The transactions for the internal `HoneyBadger` instance: this includes both user-defined
+/// "regular" transactions as well as internal transactions for coordinating node additions and
+/// removals and key generation.
+#[derive(Eq, PartialEq, Debug, Serialize, Deserialize, Hash)]
+enum Transaction<Tx, NodeUid> {
+    /// A user-defined transaction.
+    User(Tx),
+    /// A vote by an existing node to add or remove a node.
+    Change(Change<NodeUid>),
+    /// A proposal message for key generation.
+    Propose(NodeUid, Propose, Box<Signature>),
+    /// An accept message for key generation.
+    Accept(NodeUid, Accept, Box<Signature>),
+}
+
+impl<Tx, NodeUid> From<Input<Tx, NodeUid>> for Transaction<Tx, NodeUid> {
+    fn from(input: Input<Tx, NodeUid>) -> Transaction<Tx, NodeUid> {
+        match input {
+            Input::User(tx) => Transaction::User(tx),
+            Input::Change(change) => Transaction::Change(change),
+        }
+    }
+}
+
+/// A batch of transactions the algorithm has output.
+#[derive(Clone)]
+pub struct Batch<Tx, NodeUid> {
+    pub epoch: u64,
+    pub transactions: BTreeMap<NodeUid, Vec<Tx>>,
+    pub change: Option<Change<NodeUid>>,
+}
+
+impl<Tx, NodeUid: Ord> Batch<Tx, NodeUid> {
+    /// Returns a new, empty batch with the given epoch.
+    pub fn new(epoch: u64) -> Self {
+        Batch {
+            epoch,
+            transactions: BTreeMap::new(),
+            change: None,
+        }
+    }
+
+    /// Returns an iterator over all transactions included in the batch.
+    pub fn iter(&self) -> impl Iterator<Item = &Tx> {
+        self.transactions.values().flat_map(|vec| vec)
+    }
+
+    /// Returns the number of transactions in the batch (without detecting duplicates).
+    pub fn len(&self) -> usize {
+        self.transactions.values().map(Vec::len).sum()
+    }
+
+    /// Returns `true` if the batch contains no transactions.
+    pub fn is_empty(&self) -> bool {
+        self.transactions.values().all(Vec::is_empty)
+    }
+
+    /// Returns the change to the set of participating nodes, if any.
+    pub fn change(&self) -> Option<&Change<NodeUid>> {
+        self.change.as_ref()
+    }
+}
+
+/// A message sent to or received from another node's Honey Badger instance.
+#[cfg_attr(feature = "serialization-serde", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone)]
+pub enum Message<NodeUid> {
+    /// A message belonging to the `HoneyBadger` algorithm started in the given epoch.
+    HoneyBadger(u64, honey_badger::Message<NodeUid>),
```

Comment:
I suppose there are going to be other messages in this `enum`. Otherwise it would have been a newtype.

Reply:
Yes, we will need additional messages so that the joining node can take part in the key generation (DKG): It will have to send its `Propose` and `Accept`s as messages to the other nodes.

In fact, I think I'll insert this messaging step for all nodes, to prevent censorship. So every node will send its votes and DKG messages first as messages, and everyone who receives a signed message from a peer will propose it in Honey Badger. (Instead of only proposing their own messages, like they do now.)